transport-selector/src/main/java/io/netty/transport/selector/TransportSelector.java

```diff
@@ -0,0 +1,163 @@
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.transport.selector;
+
+import io.netty.channel.Channel;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.epoll.Epoll;
+import io.netty.channel.epoll.EpollDatagramChannel;
+import io.netty.channel.epoll.EpollDomainSocketChannel;
+import io.netty.channel.epoll.EpollEventLoopGroup;
+import io.netty.channel.epoll.EpollServerDomainSocketChannel;
+import io.netty.channel.epoll.EpollServerSocketChannel;
+import io.netty.channel.epoll.EpollSocketChannel;
+import io.netty.channel.kqueue.KQueue;
+import io.netty.channel.kqueue.KQueueDatagramChannel;
+import io.netty.channel.kqueue.KQueueDomainSocketChannel;
+import io.netty.channel.kqueue.KQueueEventLoopGroup;
+import io.netty.channel.kqueue.KQueueServerDomainSocketChannel;
+import io.netty.channel.kqueue.KQueueServerSocketChannel;
+import io.netty.channel.kqueue.KQueueSocketChannel;
+import io.netty.channel.nio.NioEventLoopGroup;
+import io.netty.channel.socket.DatagramChannel;
+import io.netty.channel.socket.ServerSocketChannel;
+import io.netty.channel.socket.SocketChannel;
+import io.netty.channel.socket.nio.NioDatagramChannel;
+import io.netty.channel.socket.nio.NioServerSocketChannel;
+import io.netty.channel.socket.nio.NioSocketChannel;
+import io.netty.channel.unix.DomainSocketChannel;
+import io.netty.channel.unix.ServerDomainSocketChannel;
+
+import java.util.concurrent.ThreadFactory;
+
+/**
+ * Helper class which will hand you the <strong>optimal</strong> implementation of a transport depending on the
+ * used platform and the native libraries on the classpath.
+ *
+ * It is important to note that if you use any of the methods to select the right {@link Channel} you should also use
+ * the methods provided by this class to select the right {@link EventLoopGroup}. Otherwise you may run into
+ * miss-matches between {@link Channel} and {@link EventLoopGroup} implementations.
+ */
+public final class TransportSelector {
+
+    private TransportSelector() { }
+
+    /**
+     * Returns the {@link SocketChannel} {@link Class} to use on this platform.
+     */
+    public static Class<? extends SocketChannel> socketChannel() {
+        if (Epoll.isAvailable()) {
+            return EpollSocketChannel.class;
+        }
+        if (KQueue.isAvailable()) {
+            return KQueueSocketChannel.class;
+        }
+        return NioSocketChannel.class;
```

Comment:
As I mentioned in the other issue, there are some use cases where users or applications may prefer to fail outright, rather than silently defaulting to a non-native transport. E.g., perhaps you expect Epoll to be available, and silently defaulting to Nio might have dangerous performance implications.

I'm okay with saying that is too niche of a use case for us to be concerned with here, but I'm also wondering if we should we consider offering some way to *require* a native transport? Perhaps an accompanying `NativeTransportSelector`? Or `require` variations of the methods?